### DISCUSSION TITLE
How to integrate the yourkit buildpack into a spring-boot-maven plugin image build

### DISCUSSION THREAD

**Discussion by astmuc** [paketo-buildpacks/feedback#284] [Java Team]
I try to use the yourkit buildpack with the spring-boot-maven plugin like shown at the bottom. The build runs without error and the yourkit buildpack is in use:

```
[INFO]     [creator]     Paketo Buildpack for YourKit 6.2.2
[INFO]     [creator]       https://github.com/paketo-buildpacks/yourkit
[INFO]     [creator]       YourKit Java Agent 2024.3.152: Reusing cached layer
[INFO]     [creator]       Launch Helper: Reusing cached layer
[INFO]     [creator]     ===> EXPORTING
```
My question is: What is necessary to get the javaagent option into the java option ? Right now the java command line stays as defined below.

```
      <plugin>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-maven-plugin</artifactId>
        <version>3.2.5</version>
        <executions>
          <execution>
            <phase>package</phase>
            <goals>
              <goal>repackage</goal>
              <goal>build-image-no-fork</goal>
            </goals>
          </execution>
        </executions>
        <configuration>
          <layers>
            <enabled>true</enabled>
          </layers>
          <image>
            <name>${minikubeIp}:30500/${project.artifactId}:${version}</name>
            <publish>true</publish>
            <builder>paketobuildpacks/builder-jammy-base:latest</builder>
            <buildpacks>
              <buildpack>gcr.io/paketo-buildpacks/ca-certificates</buildpack>
              <buildpack>gcr.io/paketo-buildpacks/bellsoft-liberica</buildpack>
              <buildpack>gcr.io/paketo-buildpacks/syft</buildpack>
              <buildpack>gcr.io/paketo-buildpacks/executable-jar</buildpack>
              <buildpack>gcr.io/paketo-buildpacks/dist-zip</buildpack>
              <buildpack>gcr.io/paketo-buildpacks/spring-boot</buildpack>
              <buildpack>gcr.io/paketo-buildpacks/environment-variables</buildpack>
              <buildpack>docker.io/paketobuildpacks/yourkit</buildpack>
            </buildpacks>
            <env>
              <BP_YOURKIT_ENABLED>true</BP_YOURKIT_ENABLED>
              <BPL_YOURKIT_ENABLED>true</BPL_YOURKIT_ENABLED>
              <BPL_YOURKIT_PORT>10001</BPL_YOURKIT_PORT>
              <BP_JVM_VERSION>${java.version}</BP_JVM_VERSION>
              <BPE_DELIM_JAVA_TOOL_OPTIONS xml:space="preserve"> </BPE_DELIM_JAVA_TOOL_OPTIONS>
              <BPE_APPEND_JAVA_TOOL_OPTIONS>--enable-preview -Duser.timezone=Europe/Berlin -XX:+HeapDumpOnOutOfMemoryError
                -Dreactor.schedulers.defaultBoundedElasticOnVirtualThreads=true
              </BPE_APPEND_JAVA_TOOL_OPTIONS>
            </env>
          </image>
        </configuration>
      </plugin>
```

**Comment by dmikusa** [comment]
At build time, that's all you're going to see. It just adds the agent binaries to the image, which it has done for you (actually, it's cached so not in that particular build but in a previous build you ran).

Then you need to enable it at runtime. See the [docs here for instructions on doing that](https://paketo.io/docs/howto/app-monitor/#yourkit) as well as instructions on connecting with YourKit UI.

Hope that helps!

**Reply by astmuc** [reply]
Thank your very much for the fast help !
My question was about this docu snippet of the yourkit buildpack:
```
The buildpack will do the following:

    Contributes a Java agent to a layer and configures $JAVA_TOOL_OPTIONS to use it
```
tbh i do not know how to construct a binding for the agent jar-file without it. Or is it already available at a fixed location in the image ?

When i start my app with the build above i can see the java options logged during start and there is nothing for the yourkit agent included. Means, i still miss something.

**Reply by dmikusa** [reply]
I see. I apologize as that doc statement is misleading.

YourKit Buildpack, as well as other APM tool buildpacks, have two stages. There's a build stage where you can tell it to actually contribute the agent files into the resulting image, and there's a run stage where you tell it to actually enable the agent at runtime. This comment does not differentiate between the two and that's confusing. I'll get that updated.

The doc link I sent is correct though. If you follow those instructions things should work for you.

**Comment by astmuc** [comment]
Yes, i found the location of the agent by inspecting a running pod. Means, i am profiling :-)  My first thought was, that the profile buildpack will append to the yourkit JAVA_TOOL_OPTIONS, but it seems to use a fresh start.

Thanks for your tips and the doc-link. I also corrected my pom to use urn-like buildpacks references.
